src/common/types.ts

```diff
@@ -51,7 +50,6 @@ export type CalculatedColumn<TRow, TField extends keyof TRow = keyof TRow> =
     idx: number;
     width: number;
     left: number;
-    cellContentRenderer: CellContentRenderer<TRow>;
```

Comment:
I'd rather we keep using this type for `formatter`. Making the `formatter` a function makes things so much simpler/better IMO. That way we can just do
```js
const column = {
  key: '...',
  formatter(rowData) {
    return intlFormatter.format(rowData.value);
  }
};
```

Removing the requirement for formatters to be a React component means we can reuse utils/libs directly as formatters, and we don't have to unnecessarily go through another React component instantiation/rendering.

https://gist.github.com/nstepien/090298c3c2d94324cb332c33d82fdcfb

Reply:
I did some quick profiling and did not notice any performance difference when using a component instead of a function. After the initial render, memoization should anyway prevent unnecessary cell rendering.

Here are a few reasons I prefer a component
- There is nothing that prevent us from calling component as function. For example `formatter: ImageFormatter` would work but it is technically not correct. It is fairly common to have `JSX` in a formatter and we will have to use `formatter(props) {  return <Formatter {...props} />;  }` whenever the formatter needs `JSX`. 
- Calling a component as a function sometimes works but it can lead to confusing [errors](https://kentcdodds.com/blog/dont-call-a-react-function-component), one example is hooks. We can use `formatter(props) {  return <Formatter {...props} />;  }` but I wonder if it necessary? I think it would be simple to just use `formatter: Formatter` instead. We will have to add additional documentation explaining the differences which is extra work IMO. 
- We won't catch this bug with types
- `formatter(rowData) { return intlFormatter.format(rowData.value) }` is a valid component. It does not work because of the type [issue](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20544). A workaround is `formatter(rowData) { return <>intlFormatter.format(rowData.value)</> }`

Reply:
>We will have to add additional documentation explaining the differences which is extra work IMO.

The documentation is already written for us: https://reactjs.org/docs/render-props.html

>We won't catch this bug with types

I haven't tested this, but I'd expect React to fail immediately if you use hooks outside a separate component, so tests would catch this bug instead.

Reply:
@amanmahajan7 

1. `formatter: ImageFormatter` is a little bit restricted as in `ImageFormatter`, it cannot access any outside scope data and it can only rely on `redux` or `context`, which actually makes things more difficult, so for this usage, I would actually want to gradually deprecate it. When we pass an element directly, it will result in the issue that I mentioned in the previous comment in some cases.
1.1 Let's say if `ImageFormatter` is a class component, what we do inside the data grid is to check whether it's a valid component and then render it, It's same if the user provides the formatter function and just do `formatter(props) { return <ImageFormatter/>; }`
1.2 If the `ImageFormatter` is a function component without hook, it can just be used as a function.
1.3 If the `ImageFormatter` is a function component with hook. it falls to 1.1

> We won't catch this bug with types
Could you be more specific? I think it's the user's decision to decide what's being rendered in the cell, By default we only need to support `row[column.key]`, but they can even only display some hard coded text based on their business requirement.

Reply:
Render prop is mainly for sharing common logic but we are using it to improve performance but I did not notice any difference in my testing . With component we can prevent any edge cases so I am not sure what is the advantage of using render props in this case. 

> I haven't tested this, but I'd expect React to fail immediately if you use hooks outside a separate component, so tests would catch this bug instead.

This would be a run time error which can totally be avoided with component

> We won't catch this bug with types
I meant the following would be valid with render props and there would not be any type error

```
formatter(props) {
  const xxx = useState(someDataInMyGrid);
  return someElement;
}
```